Scripts/Silverpine: fixed Lordaeron (questId 27098)

Nyr97 opened the pull request:
Please, do not attempt to merge this PR until https://github.com/TrinityCore/TrinityCore/pull/28275, https://github.com/TrinityCore/TrinityCore/pull/28303, and https://github.com/TrinityCore/TrinityCore/pull/28334 are merged since this one depends on those three.

**Changes proposed:**

* Scripted RP event for quest Lordaeron (questId 27098).
* Quest fixed:

- [x] Fixed quest Lordaeron (questId 27098).

**Issues addressed:**

None.

**Tests performed:**
 
It builds and it was tested in-game.

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Currently, the player gets dismounted from their horse when the vehicle uses SPELL_RIDE_SYLVANAS_HORSE to mount on Sylvanas' horse seatId 1. This issue is triggered by how this type of summonProperty (vehicle) is handled: as of today, any vehicle summoned creature will force Hardcoded Ride Vehicle. However, auras with control effect should respect its base value if higher than 10 because it references a different control effect spell, so it should be casting that one instead. TL;DR SummonProperties.db2 needs to be implemented properly.

Nyr97 committed:
Scripts/Silverpine: fixed Lordaeron (questId 27098).

Nyr97 committed:
fix build

Nyr97 committed:
add correct speed for pathing

Nyr97 committed:
build fix x2

Nyr97 committed:
build fix

Nyr97 committed:
removed unnused variables and fixed an infinite loop

offl commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1170,6 +1178,26 @@ struct npc_silverpine_sylvanas_windrunner_high_command : public ScriptedAI
+                player->CastSpell(player, SPELL_SUMMON_SYLVANAS_AND_HORSE, true);
+                player->CastSpell(player, SPELL_SUMMON_FORSAKEN_WARHORSE, true);
+                player->CastSpell(player, SPELL_LORDAERON_AURA, true);
+                player->CastSpell(player, SPELL_SUMMON_LORDAERON_ACTORS, true);
```
>```cpp
>    SPELL_LORDAERON_AURA                        = 84189, // Note: NYI, after researching thoroughly, nothing seems to point what this SPELL_AURA_DUMMY should do. Maybe it is what forces the player to cast the rest of the spells.
>```

I'm almost sure SourceSpellID (spell to cast on quest accept) is a thing on retail, despite it's not sent to client because not needed by client. That is easiest way to cast spells comparing to other possible ways. But what is done here I have no idea. Scripted auras for such stuff usually has small duration, this one infinite and if you preserved orded of casts, it does something else. All those spells also are casted by player, so maybe there's serverside master spell

Btw what happens if quest is abandoned and taken again? Player casts again all those spells? Using SourceSpellID raises question how to add condition for [this](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=20130), which is much more easier case
When this aura gets removed btw?

Nyr97 replied:
I checked the serverside spell list that we have access to and there's no master spell related to any of these at all. I checked around 250 ids before and after these, and there's nothing. I need to review the packets being sent because right now I'm not sure if I really preserved the order for this exact spell.

If the quest is retaken, all of the spells are recast again. The aura gets removed either when the player dismounts from the horse, which causes a skip to happen, and then the given complete quest spell is cast, or after that complete quest spell is cast.

Nyr97 replied:
I just rechecked the packets and they follow this order in the same world tick:

SMSG_SPELL_START - 84128 (Summon Sylvanas + Horse)
SMSG_SPELL_START - 84164 (Summon Forsaken Warhorse)
SMSG_SPELL_START - 85406 (Detect: Quest Invis Zone 20 (Ambermill Shield)) [This one is handled by area_spell]
SMSG_SPELL_START - 84127 (Summon Lordaeron Actors)
SMSG_SPELL_START - 84189 (Lordaeron Area Aura)
SMSG_SPELL_START - 46598 (Ride Vehicle Hardcoded)
SMSG_SPELL_GO - 84189 (Lordaeron Area Aura)
SMSG_SPELL_GO - 84128 (Summon Sylvanas + Horse)
SMSG_SPELL_GO - 85406 (Detect: Quest Invis Zone 20 (Ambermill Shield))
SMSG_SPELL_GO - 84164 (Summon Forsaken Warhorse)
SMSG_SPELL_GO - 84127 (Summon Lordaeron Actors)
SMSG_SPELL_GO - 46598 (Ride Vehicle Hardcoded)

offl replied:
It's exactly what I usually see, master spell is sent as last

Nyr97 replied:
So are we to assume that aura spell dummy is the masterspell for everything else?

offl replied:
>The aura gets removed either when the player dismounts from the horse, which causes a skip to happen, and then the given complete quest spell is cast, or after that complete quest spell is cast.

So it serves some kind of purpose, because of that it should be an aura with infinite duration

Nyr97 replied:
Feedback applied!

Nyr97 replied:
I've done more research on the packets, apparently, the aura is first applied and cast when No Escape quest is rewarded. It is on until Lordaeron quest is rewarded, so it cannot be the master spell for all of the rest.

offl replied:
Any way, I have no idea how to use SourceSpellID for player cast, didn't encountered something like that in 335 yet

Nyr97 replied:
SourceSpellId in master is contained in quest_template_addon, but if I remember correctly that value is taken directly from the sniff if there's any difference between the sniff and the DB and I had no update row.

offl commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1624,714 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+};
+
+// 84173 - Despawn All Summons
+class spell_silverpine_despawn_all_summons_lordaeron : public SpellScript
```
Maybe create a generic spell script, there's like 100 similar spells, spell_gen_despawn_owned or something like that

Nyr97 replied:
I'll consider it, but I'm most likely taking that approach since we can control target conditions through DB.

Nyr97 replied:
I have added this spell as generic in an older PR which is soon getting merged.

Nyr97 committed:
moved Despawn All Summons to a generic spell

* also moved a spell cast to its appropriate placement according to packets.

Nyr97 committed:
Revert "moved Despawn All Summons to a generic spell"

This reverts commit fcd89615161b5d2a35840e4055bf953eb2c28d23.

Nyr97 committed:
handled Lordaeron Area Aura as master spell for every spell cast during Lordaeron quest

mdX7 commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1618,654 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+        _playerGUID = summoner->GetGUID();
+    }
+
+    void EnterEvadeMode(EvadeReason /*why*/) override { }
```
add a comment why you are overriding it like this, looks kinda unused on first look although its serving a special purpose

Nyr97 replied:
Feedback applied. It was all for testing purposes, it's now gone.

mdX7 commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1618,654 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+            {
+                case EVENT_ACTIVATE_SKIP:
+                {
+                    if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
```
just use `if (Unit* summoner = me->GetSummonerUnit()` its doing all the checks by itself, you can drop `_playerGUID` in the process

Nyr97 replied:
Feedback applied though I need to access ToTempSummon() for this.

mdX7 commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1618,654 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+};
+
+// 45057 - Forsaken Warhorse (Sylvanas)
+struct npc_silverpine_warhorse_sylvanas_lordaeron : public ScriptedAI
```
looks like a script which could be moved to sai

Nyr97 replied:
I really want to avoid moving anything related to sAI when it comes to RP events, especially because the mounting bug is still existing.

mdX7 commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1618,654 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+                {
+                    me->HandleEmoteCommand(EMOTE_STATE_HOLD_BOW);
+
+                    if (caster->ToUnit()->GetEntry() == NPC_WORGEN_RENEGADE)
```
`ToUnit()` cast is not required; `GetEntry()` is part of `WorldObject`

Nyr97 replied:
Feedback applied, in 434 branch GetEntry was only part of Unit/Creature.

mdX7 commented on src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:
```diff
@@ -1596,6 +1618,654 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+
+    void Reset() override
+    {
+        _events.Reset();
```
also set `_done = false;` here

Nyr97 replied:
Feedback applied.

Nyr97 committed:
feedback applied

Nyr97 committed:
changes to Lordaeron Area Aura

the aura should be gone on quest completion, which also removes SPELL_SUMMON_FORSAKEN_WARHORSE from the player (it adds an invisibility mask).

Nyr97 committed:
small updates

* added missing visibilityDistanceType for certain creatures.
* added Lordaeron Area Aura with correct implementation.
* added correct UnitFlags to some creatures.
* refactored SPELL_KILL_ME on SpellHit.

mdX7 committed:
Merge branch 'master' into teleqraph/Silverpine-Lordaeron

mdX7 committed:
Merge remote-tracking branch 'remotes/origin/master' into teleqraph/Silverpine-Lordaeron

mdX7 committed:
fix merge fail

Nyr97 committed:
small updates

Nyr97 committed:
improved Sylvanas' code

Nyr97 committed:
fixed a bugged logic on Sylvanas

Nyr97 committed:
code improvements

Nyr97 committed:
small update

Nyr97 committed:
added missing condition target for ride spell

Nyr97 committed:
added skip back to Forsaken Warhorse

mdX7 committed:
Merge branch 'master' into teleqraph/Silverpine-Lordaeron

Nyr97 committed:
update sql

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into Silverpine-Lordaeron

mdX7 committed:
Merge remote-tracking branch 'remotes/origin/master' into teleqraph/Silverpine-Lordaeron

Nyr97 committed:
SQL: fix build

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into Silverpine-Lordaeron
